awx/main/tasks.py

```diff
@@ -1482,38 +1479,85 @@ def run(self, pk, **kwargs):
+            # reading.
+            sockin, sockout = socket.socketpair()
+
+            # Spawned in a thread so Receptor can start reading before we finish writing, we
+            # write our payload to the left side of our socketpair.
+            def transmit(_socket):
+                ansible_runner.interface.run(streamer='transmit',
+                                             _output=_socket.makefile('wb'),
+                                             **params)
+
+                # Socket must be shutdown here, or the reader will hang forever.
+                _socket.shutdown(socket.SHUT_WR)
+
+            threading.Thread(target=transmit, args=[sockin]).start()
+
+            # We establish a connection to the Receptor socket and submit our work, passing
+            # in the right side of our socketpair for reading.
+            receptor_ctl = ReceptorControl('/var/run/receptor/receptor.sock')
+            result = receptor_ctl.submit_work(worktype=worktype,
+                                              payload=sockout.makefile('rb'))
+            sockin.close()
+            sockout.close()
+
+            self.update_model(self.instance.pk, receptor_work_id=result['unitid'])
+
+            resultsock, resultfile = receptor_ctl.get_work_results(result['unitid'],
+                                                                   return_socket=True,
+                                                                   return_sockfile=True)
+
+            def processor():
+                return ansible_runner.interface.run(streamer='process',
+                                                    quiet=True,
+                                                    _input=resultfile,
+                                                    event_handler=self.event_handler,
+                                                    finished_callback=self.finished_callback,
+                                                    status_handler=self.status_handler)
+
+            def cancel_watcher(processor_future):
+                while True:
+                    if processor_future.done():
+                        return
+
+                    if self.cancel_callback():
+                        result = namedtuple('result', ['status', 'rc'])
+                        return result('canceled', 1)
+                    time.sleep(1)
+
+            # Both "processor" and "cancel_watcher" are spawned in separate threads.
+            # We wait for the first one to return. If cancel_watcher returns first,
+            # we yank the socket out from underneath the processor, which will cause it
+            # to exit. A reference to the processor_future is passed into the cancel_watcher_future,
+            # Which exits if the job has finished normally. The context manager ensures we do not
+            # leave any threads laying around.
+            with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
```

Comment:
This is some fancy new python3 stuff, and I think it's a reasonable way to accomplish this, but we'll want to pay careful attention to integration results to make sure we don't notice any sort of weird race conditions/hangs related to our logging code and this threadpool.  This sort of thing has bitten us in the past due to cpython bugs:

https://bugs.python.org/issue37429

Reply:
Would you feel better with using `ProcessPoolExecutor`?

Reply:
Nope (for reasons we privately discussed) 😄